hare3/protocol.go

```diff
@@ -0,0 +1,397 @@
+			out.message = &Message{Body: Body{
+				IterRound: p.IterRound,
+				Value:     Value{Reference: ref},
+			}}
+		}
+	}
+}
+
+func (p *protocol) next(active bool) output {
+	out := output{}
+	p.execution(&out, active)
+	if p.Round >= softlock && p.coin != nil && !p.coinout {
+		coin := p.coin.LSB() != 0
+		out.coin = &coin
+		p.coinout = true
+	}
+	if p.Round == preround && p.Iter == 0 {
+		// skips hardlock unlike softlock in the paper.
+		// this makes no practical difference from correctness.
+		// but allows to simplify assignment in validValues
+		p.Round = softlock
+	} else if p.Round == notify {
+		p.Round = hardlock
+		p.Iter++
+	} else {
+		p.Round++
+	}
+	return out
+}
+
+type gossipInput struct {
+	*input
+	received      IterRound
+	otherReceived *IterRound
+}
+
+// Protocol 1. graded-gossip. page 10.
+type gossip struct {
+	threshold uint16
+	state     map[messageKey]*gossipInput
+}
+
+func (g *gossip) receive(current IterRound, input *input) (bool, *types.HareProof) {
+	// Case 1: will be discarded earlier
+	other, exist := g.state[input.key()]
+	if exist {
+		if other.msgHash != input.msgHash && !other.malicious {
+			// Case 3
+			other.malicious = true
+			other.otherReceived = &current
+			return true, &types.HareProof{Messages: [2]types.HareProofMsg{
+				other.ToMalfeasanceProof(), input.ToMalfeasanceProof(),
+			}}
+		}
+		// Case 2. but also we filter duplicates from p2p layer here
+		return false, nil
+	}
+	// Case 4
+	g.state[input.key()] = &gossipInput{input: input, received: current}
+	return true, nil
+}
+
+type gset struct {
+	values   []types.ProposalID
+	grade    grade
+	smallest types.VrfSignature
+}
+
+// Protocol 2. gradecast. page 13.
+func (g *gossip) gradecast(target IterRound) []gset {
+	// unlike paper we use 5-graded gossip for gradecast as well
+	var rst []gset
+	for key, value := range g.state {
+		if key.IterRound == target && !value.malicious {
+			g := grade0
+			if value.atxgrade == grade5 && value.received.Delay(target) <= 1 &&
+				// 2 (a)
+				(value.otherReceived == nil || value.otherReceived.Delay(target) > 3) {
+				// 2 (b)
+				g = grade2
+			} else if value.atxgrade >= grade4 && value.received.Delay(target) <= 2 &&
+				// 3 (a)
+				(value.otherReceived == nil || value.otherReceived.Delay(target) > 2) {
+				// 3 (b)
+				g = grade1
+			}
+			if g > grade0 {
+				rst = append(rst, gset{
+					grade:    g,
+					values:   value.Value.Proposals,
+					smallest: value.Eligibility.Proof,
+				})
+			}
+		}
+	}
+	// it satisfies p-Weak leader election
+	// inconsistent order of proposals may cause participants to commit on different proposals
+	sort.Slice(rst, func(i, j int) bool {
+		return rst[i].smallest.Cmp(&rst[j].smallest) == -1
+	})
+	return rst
+}
+
+// Protocol 3. thresh-gossip. Page 15.
+// output returns union of sorted proposals received
+// in the given round with minimal specified grade.
+func (g *gossip) thresholdGossip(filter IterRound, grade grade) []types.ProposalID {
```

Comment:
sorry i don't follow

more practically speaking, lets use preround messages as an example (softlock is skipped):
- at hardlock votes are computed from min(atx grade, msg grade) == 5
- at propose votes are computed from min(atx grade, msg grade) >= 4
- at wait1 votes are computed from min(atx grade, msg grade) >= 3
- at wait2 votes are computed from min(atx grade, msg grade) >= 2

note that msg grade is implicit because of the timing when thresholdGossip is called. e.g at hardlock we will receive preround messages within single delay. at propose - within 2 delays and so on.

then we ensure that votes crossed threshold, and there is atleast one honest vote.

Reply:
the same for commit:

notify: min(atx grade, msg grade) >= 5
hardlock: min(atx grade, msg grade) >= 4
softlock: min(atx grade, msg grade) >= 3
propose: min(atx grade, msg grade) >= 2
commit: min(atx grade, msg grade) >= 1

same caveat about msg grade, it is implicit because when threshold gossip is called.

i also rechecked this part of the protocol:
- valid messages do not include malicious, they are very clearly separate, the protocol checks that there is atleast one non-malicious. see 4th in 3(a)
- filtering in 3(a) is applied almost as is in the code, filter by round (1st), filter by grade (3rd, 4th)
- then we count malicious messages, why would they be a part of valid messages?
- and threshold is passed if there is atleast 1 valid/good message, and total number of messages (together) with equivocation crossed threshold